Introducing the bigint

mununki opened the pull request:
Discussions: #6525 

This PR is to introduce the JS [bigint](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) as a primitive type. With the addition of the bigint type, this PR implements support for basic arithmetic operators and unary negation operators as a first step. The comparison operator and exponential operator is being added to the initial implementation of this PR for review.

The main changes are

1. Changed the deprecated nativeint to bigint: use the `*n` literal of nativeint for bigint.
2. Added arithmetic operators. (The comma can be changed to something better on the discussion. I initially considered `+n`, but saw the increased complexity in parsing and implemented it as a comma).
```
1n +, 1n // add
1n -, 1n // sub
1n *, 1n // mul
1n /, 1n // div
```
3. Omitted optimizations at build time because there is no API in OCaml that supports bigint-sized numbers, unlike int and float

cometkim commented on jscomp/others/js_json.res:
```diff
@@ -30,6 +30,7 @@ type rec t =
   | @as(null) Null
   | String(string)
   | Number(float)
+  | Bigint(bigint)
```
BigInt doesn't belong to JSON spec

mununki replied:
Ah, correct. I'll remove it, thanks!

mununki replied:
Removed https://github.com/rescript-lang/rescript-compiler/commit/c8b78d26d225c06e2d04be9c6ba5f4837259e638

zth commented:
Can be done in a follow up PR of course, but adding support for it in untagged variants would be nice.

mununki commented:
> Can be done in a follow up PR of course, but adding support for it in untagged variants would be nice.

Oh, this PR already added support the untagged variant with Bigint.

zth commented:
@mununki haha I'm sorry, you're right... Well done!

zth commented on jscomp/ml/ast_payload.ml:
```diff
@@ -82,6 +82,20 @@ let is_single_float (x : t) : string option =
+              ({pexp_desc = Pexp_constant (Pconst_integer (name, Some n)); _}, _);
+          _;
+        };
+      ] when n = 'n' ->
```
Nit, but can't this when clause just be directly inside of the pattern?

zth commented on jscomp/ml/ast_untagged_variants.ml:
```diff
@@ -375,7 +393,7 @@ module DynamicChecks = struct
     let literals_overlaps_with_number () =
       Ext_list.exists literal_cases (function
-        | Int _ | Float _ -> true
+        | Int _ | Float _ | Bigint _ -> true
```
Bigint doesn't overlap with number, right?

mununki replied:
Thanks! https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/a9e4c162004da2b3e3f4a023e67f19c75138380f

zth commented on jscomp/others/js_bigint.res:
```diff
@@ -1,3 +1,11 @@
 /*** JavaScript BigInt API */
 
-type t
+type t = bigint
```
Considering whether we should just remove this one and refer to `bigint` directly instead. It's breaking so we could wait for a bit before doing that and just deprecate `t` meanwhile, but we should definitely refer to `bigint` instead of `t` in the code below.

mununki replied:
Thanks! https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/cf62db8067252acebd6eba665885b22080997bbc

zth commented on jscomp/others/js_bigint.res:
```diff
@@ -1,3 +1,11 @@
-type t
+type t = bigint
+
+external \"~-": t => t = "%negbigint"
+external \"~+": t => t = "%identity"
+external \"+": (t, t) => t = "%addbigint"
+external \"-": (t, t) => t = "%subbigint"
+external \"*": (t, t) => t = "%mulbigint"
+external \"/": (t, t) => t = "%divbigint"
+external mod: (t, t) => t = "%modbigint"
```
As said above, I think `t` should be `bigint` instead here.

zth reviewed (commented):
Great work! 🎉 This will be great to support. A few smaller questions from me. @cristianoc should have a look at this as well.

I'm not sold on the comma. Comma has such a specific meaning in the language that the code looks broken to me when there's a comma. I don't have a better suggestion yet though, but maybe we could figure one out together.

cristianoc commented:
Having yet another syntax for one more numeric type seems to make an existing problem worse.

The existing problem is having to use + and +. for ints and floats.

I'd rather go the direction of making x+y resolve to int or float or bigint at type checking time.

zth commented:
While we figure something like that out, can we get away with the following perhaps?
```rescript
// Using opens, described in this PR
open! Js.Bigint

let a = 1n + 1n

// Or, functions on bigint...?
let a = Js.Bigint.add(1n, 1n)
```

cknitt commented:
> I'd rather go the direction of making x+y resolve to int or float or bigint at type checking time.

This would be great, but how would it work with type inference? What type would the function

```rescript
let f = (x, y) => x + y
```

have?

cristianoc commented:
> > I'd rather go the direction of making x+y resolve to int or float or bigint at type checking time.
> 
> 
> 
> This would be great, but how would it work with type inference? What type would the function
> 
> 
> 
> ```rescript
> 
> let f = (x, y) => x + y
> 
> ```
> 
> 
> 
> have?

Without type annotations, the default numeric type.
One could define a generic numeric type, though it's probably overkill as polymorphic numerical code is not very likely.

cometkim commented:
> > I'd rather go the direction of making x+y resolve to int or float or bigint at type checking time.
>
> This would be great, but how would it work with type inference? What type would the function

suggested on #6477 

actually we are doing same for comparison operators.

mununki commented:
I agree adding another number type to make a problem worse. I didn't think about adding the generic number type yet. I think it is very good direction to solve the problem having two different operators for int and float. I grab a chance to look into how int and float type value are handled during compilation while working on this PR. Actually, I couldn't find many reasons to separate two different type(int, float) for emitting the JS output except optimization of int32 for int type value.

How about defining the `number` type and deprecating the int, float type in future?(maybe we can come back and decide the deprecation later) I can start working on defininig the number type and simplifying the operators with another PR. How about it?

cristianoc commented:
I suspect a number of users rely on the semantic difference between int and float. Eg with float it's easy to get 0.999999 while 1 is expected so simple equality does not work.

That's why I was suggesting a simple change at the level of type checker, which does not break existing code.
Just, decides at type checking time what number type is intended.

mununki commented:
I found I missed the coercion for bigint, then I've added the commits https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/b13884d2b5f7eda6e12fc339600480a414aeceb7 @zth

mununki commented:
> I suspect a number of users rely on the semantic difference between int and float. Eg with float it's easy to get 0.999999 while 1 is expected so simple equality does not work.
> 
> That's why I was suggesting a simple change at the level of type checker, which does not break existing code. Just, decides at type checking time what number type is intended.

If I understand correctly, Do you mean letting the type checker infer types for arithmetic operations like it does for comparisons?

```ocaml
// Int operations
external \"+": (int, int) => int = "%addint"
// Float operations
external \"+.": (float, float) => float = "%addfloat"

// to

// Number operations
external \"+": ('a, 'a) => 'a = "%addnumber"
```

cristianoc commented:
> > I suspect a number of users rely on the semantic difference between int and float. Eg with float it's easy to get 0.999999 while 1 is expected so simple equality does not work.
> 
> > 
> 
> > That's why I was suggesting a simple change at the level of type checker, which does not break existing code. Just, decides at type checking time what number type is intended.
> 
> 
> 
> If I understand correctly, Do you mean letting the type checker infer types for arithmetic operations like it does for comparisons?
> 
> 
> 
> ```ocaml
> 
> // Int operations
> 
> external \"+": (int, int) => int = "%addint"
> 
> // Float operations
> 
> external \"+.": (float, float) => float = "%addfloat"
> 
> 
> 
> // to
> 
> 
> 
> // Number operations
> 
> external \"+": ('a, 'a) => 'a = "%addnumber"
> 
> ```

Yes a bit like for comparison.
Except while fully polymorphic comparison exists, fully polymorphic addition does not.
One can't have a single function that works for both int and float, but needs to generate one or the other.

mununki commented:
To summarize, if we'll discuss the issue of unifying the int and float operators separately, and if everyone agrees that `open! Bigint` and make the int operator available to bigint as well, I'm going to add commits for the exponential operator `**` and additional optimizations to the comparison operator in this PR.
```res
// exponentiation
let x = 1n ** 1n

// comparison optimization
let y = 1n == 1n
// compiled to var y = true;
```

zth commented on lib/es6/caml.js:
```diff
@@ -39,6 +39,20 @@ function float_compare(x, y) {
+    return 0;
+  } else if (x < y) {
+    return -1;
+  } else if (x > y || x === x) {
```
Weird or clause

mununki replied:
Oh good catch, I didn't notice because I was using the float_compare code as is. Why does float_compare have this branching as opposed to int_compare? Very weird to me too.

mununki replied:
I think those conditions are for below test cases;
```res
// float_test.res
eq(__LOC__, float_compare(Js.Float._NaN, Js.Float._NaN), 0)
eq(__LOC__, generic_compare(Js.Float._NaN, Js.Float._NaN), 0)
eq(__LOC__, float_compare(Js.Float._NaN, neg_infinity), -1)
eq(__LOC__, generic_compare(Js.Float._NaN, neg_infinity), -1)
```

mununki replied:
I don't know why float_compare was implemented this way at the time, but my best guess is that it was to capitalize on the fact that comparing NaNs is false.
```js
-Infinity === -Infinity // true
NaN === NaN // false
```
However, the bigint doesn't have a binding to Infinity or Nan for now. I'm going to fix `bigint_compare` as `int_compare`.

mununki replied:
생각을 바꿨습니다. `Js.Bigint`에도 `NaN`과 method 들에 대한 interop을 추가하고,`bigint_test.res` 테스트도 추가해보았습니다. 한 번 봐주시겠어요? https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/c972a1374643aaf566b544c41c89ba5fa02c146e

mununki replied:
I changed my mind. I've added interop for `NaN` and methods to Js.Bigint, and also added the `bigint_test.res` test. Can you take a look? https://github.com/rescript-lang/rescript-compiler/commit/c972a1374643aaf566b544c41c89ba5fa02c146e

mununki replied:
I can not say that the implementation of `float_compare` is the best but it is practical and working.

cometkim replied:
NaN is a concept that belongs to Number (`typeof NaN === `number`), not BigInt. BigInt operations never produce NaN AFAIK

mununki replied:
That is the compare function. Those clauses are for the comparing with NaN.

mununki replied:
I got your point. NaN can not be produced with BigInt operation. NaN is produced only below 5 cases:
> There are five different types of operations that return NaN:
>
> * Failed number conversion (e.g. explicit ones like parseInt("blabla"), Number(undefined), or implicit ones like Math.abs(undefined))
> * Math operation where the result is not a real number (e.g. Math.sqrt(-1))
> * Indeterminate form (e.g. 0 * Infinity, 1 ** Infinity, Infinity / Infinity, Infinity - Infinity)
> * A method or expression whose operand is or gets coerced to NaN (e.g. 7 ** NaN, 7 * "blabla") — this means NaN is contagious
> * Other cases where an invalid value is to be represented as a number (e.g. an invalid [Date](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date) new Date("blabla").getTime(), "".charCodeAt(1))

mununki replied:
So wouldn't it be better to get rid of NaNs in Js_bigint as well?

cometkim replied:
Yes. NaN is obviously number type. Date, parseInt, Math are all about Number, not BigInt.

mununki replied:
Thanks!

https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/6601152d3b720a7fdf162c2c22807d5a2ced29ac
https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/ea49c0534c92ca4e552c8fbca1c9d533d44a4504

mununki commented:
While I'm working on the comparison lambda, lam, I think I found there are some codes better to be cleaned up. E.g. the physical comparison (`%eq`, `%noteq`) is converted by `int_comp` in `js_exp_make.ml`. However, the case of `%notequal` is being omitted from the optimization, and there seems to be some ambiguity in the function name. I'll open an issue and work on it in a separate PR.

mununki commented:
I've finished adding the comparison operation optimization and exponetiation operator `**,` for bigint. I have nothing left for this PR now.

cometkim commented on jscomp/others/js_bigint.res:
```diff
@@ -1,3 +1,98 @@
+therefore necessary to test for `_NaN`. Return `true` if the given value is
+`_NaN`, `false` otherwise. See [`isNaN`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isNaN) on MDN.
+*/
+external isNaN: bigint => bool = "isNaN"
```
I don't think this is necessary in bigint module. Unlike number, bigint operations (except comparison) cannot be mixed with other types and never produce NaN.

mununki replied:
It makes sense to me.

cometkim commented on jscomp/others/js_bigint.res:
````diff
@@ -1,3 +1,98 @@
+Js.Bigint.isFinite(1234.)
+```
+*/
+external isFinite: bigint => bool = "isFinite"
````
Same here. Js BigInt never produces Infinity

cometkim commented on jscomp/ml/translcore.ml:
```diff
@@ -350,6 +365,19 @@ let primitives_table =
       ("%lefloat", Pfloatcomp Cle);
       ("%gtfloat", Pfloatcomp Cgt);
       ("%gefloat", Pfloatcomp Cge);
+      ("%negbigint", Pnegbigint);
+      ("%addbigint", Paddbigint);
+      ("%subbigint", Psubbigint);
+      ("%mulbigint", Pmulbigint);
+      ("%divbigint", Pdivbigint Safe);
+      ("%powbigint", Ppowbigint);
+      ("%modbigint", Pmodbigint Safe);
+      ("%eqbigint", Pbigintcomp Ceq);
+      ("%noteqbigint", Pbigintcomp Cneq);
+      ("%ltbigint", Pbigintcomp Clt);
+      ("%lebigint", Pbigintcomp Cle);
+      ("%gtbigint", Pbigintcomp Cgt);
+      ("%gebigint", Pbigintcomp Cge);
```
It seems `%lslbigint`(`<<`), `%lsrbigint`(`>>`), `%andbigint`(`&`), `%orbigint`(`|`), `%xorbigint`(`^`) are missing? (not sure bitwise not op) Logical operators are the main interface of BigInt.

Note that ASL(Arithmetic Shift Right, aka Unsigned right shift in JS) doesn't accept BigInt.

mununki replied:
Initially, I wasn't going to include the bitwise operation implementation in this PR because I was concerned that it would make the PR too large, but now that I think about it, it doesn't make much difference, so I'll add it.

mununki replied:
Added https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/9060aa0e4fe40aafd7ca0605c6e6603890093d37

zth commented:
I'm a bit short on time at the moment so I can't do a thorough review right now. Maybe @cristianoc can have a final look?

Also, since we haven't managed to get 11.1 out the door yet one thought I had is whether we should ship this in 11.1 already perhaps?

mununki commented:
> I'm a bit short on time at the moment so I can't do a thorough review right now. Maybe @cristianoc can have a final look?
> 
> Also, since we haven't managed to get 11.1 out the door yet one thought I had is whether we should ship this in 11.1 already perhaps?

Once you've decided which version you want to ship to, I'll rebase and update the change log.

cristianoc commented:
Looks like there are still changes in the parser for operators like `+,` while the discussion is ongoing.
How about remove those syntax extensions for binary operators.

Also, perhaps check that there is consensus on the notation `1n` for literals? Seems fine to me, but it does not seem to have been discussed.

mununki commented:
> Looks like there are still changes in the parser for operators like `+,` while the discussion is ongoing. How about remove those syntax extensions for binary operators.

I agree. As long as you declare and use `open! Js.Bigint` and using it, I don't think it's a problem. I think we only need to remove that operators from `Pervasives*.res`. What do you think?
> Also, perhaps check that there is consensus on the notation `1n` for literals? Seems fine to me, but it does not seem to have been discussed.

The BigInt literal in Js is the same to `1n`, So I hope there's much to disagree with.

cristianoc commented:
> > Looks like there are still changes in the parser for operators like `+,` while the discussion is ongoing. How about remove those syntax extensions for binary operators.
> 
> I agree. As long as user use `open! Js.Bigint`, I don't think it's a problem. I think we only need to remove that operators from `Pervasives*.res`. What do you think?

Yes, and also revert the changes in the parser.

> > Also, perhaps check that there is consensus on the notation `1n` for literals? Seems fine to me, but it does not seem to have been discussed.
> 
> The BigInt literal in Js is the same to `1n`, So I hope there's much to disagree with.
great!

zth commented:
Yeah `1n` is the right way imo.

mununki commented:
The operators for bigint are removed from parser and Pervasives now.

zth commented:
@mununki let's go for v11.1! Would you rebase to `11.0_release`?

cristianoc commented on jscomp/core/js_exp_make.ml:
```diff
@@ -1253,6 +1260,18 @@ let rec int32_band ?comment (e1 : J.expression) (e2 : J.expression) :
+  match (cmp, e0.expression_desc, e1.expression_desc) with
+  | Ceq, Number (Bigint {i = i0; _}), Number (Bigint {i = i1; _}) -> bool (i0 = i1)
+  | Cneq, Number (Bigint {i = i0; _}), Number (Bigint {i = i1; _}) -> bool (i0 <> i1)
+  | Cge, Number (Bigint {i = i0; _}), Number (Bigint {i = i1; _}) -> bool (i0 >= i1)
```
these are string literals right? then `>=` etc won't work

mununki replied:
https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/acebd8fa804f92afd6a5a21dbbf4439feecefb53

cristianoc commented on jscomp/core/lam_compat.ml:
```diff
@@ -68,15 +68,15 @@ let cmp_nativeint (cmp : comparison) (a : nativeint) b : bool =
   match cmp with
   | Ceq -> a = b
   | Cneq -> a <> b
   | Cgt -> a > b
```
these are not going to work for string right?

mununki replied:
Oh, I think I missed it, you're correct.

mununki replied:
https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/acebd8fa804f92afd6a5a21dbbf4439feecefb53

cristianoc commented on jscomp/ml/parmatch.ml:
```diff
@@ -264,6 +264,8 @@ let const_compare x y =
   | Const_float f1, Const_float f2 ->
       compare (float_of_string f1) (float_of_string f2)
+  | Const_bigint b1, Const_bigint b2 ->
+      String.compare b1 b2
```
Looks like it's only used in this form `const_compare c1 c2 = 0`, which should be mostly OK.
Except at most `010n` vs `10n` which ends up being not equal.

mununki replied:
Good catch! `01n` This case should be handled. At least it should be printed `1n` in Js output, or build error.

cristianoc commented on jscomp/ml/printlambda.ml:
```diff
@@ -177,6 +177,26 @@ let primitive ppf = function
+  | Plslbigint -> fprintf ppf "lsl"
+  | Pasrbigint -> fprintf ppf "asr"
+  | Pdivbigint Safe -> fprintf ppf "/n"
+  | Pdivbigint Unsafe -> fprintf ppf "/nu"
```
are these safe and unsafe operations supported in the current PR?

Also, what about `/n` and `/nu`: does this require binary operators, or how else are these triggered?

mununki replied:
I'm working on it.

cristianoc commented on jscomp/syntax/tests/printer/other/case.res:
```diff
@@ -93,7 +93,8 @@ let precedence = x => switch x {
   | ForwardslashDot => 6
-  | Exponentiation => 7
+  | Exponentiation
+  | ExponentiationComma => 7
```
leftover

mununki replied:
Oops, removed https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/ea0bc090f5833a7cc9574773f61a26981c7c7639

cristianoc commented on jscomp/test/VariantCoercion.res:
```diff
@@ -80,3 +80,15 @@ module CoerceFromFloatToVariant = {
 }
+
+module CoerceFromBigintToVariant = {
+  @unboxed type bigints = Bigint(bigint) | @as(1n) First | @as(2n) Second | @as(3n) Third
```
this should give an error as `Bigint(bigint)` overlaps with ` @as(1n) First`.

zth replied:
Should they really? I believe the logic we have for other types is that it's not considered overlap when concrete values are used, as they're checked before type of etc.

mununki replied:
Since variant for bigint doesn't have any special handling, it's the same structure as the coercion test for int or float, so I don't think it should throw an error.

cristianoc replied:
I think I was misremembering. No problem here.

cristianoc reviewed (commented):
OK so any leading `0` such as `010n` should be a syntax error, as it't not valid in JS.

Then, thinking about it, it looks like static comparison for bigint literals can be done by:

1) comparing size
2) if equal, comparing as string

So `100n` > `99n` because of size. And `101n` > `100n` because of string comparison.

cometkim commented:
> So 100n > 99n because of size. And 101n > 100n because of string comparison.

It need more conditions for negative bigints (e.g. `-100n > 99n` should be false)

mununki commented:
There is another case to be considered. e.g.. `0x11n > 0b10n` should be true. At the moment I don't have a good idea how to compare in this case, do you have any ideas?

cristianoc commented:
If pattern matching for bigint literals is supported, perhaps add one test like `| 1n => ...`

cristianoc commented:
> There is another case to be considered. e.g.. `0x11n > 0b10n` should be true. At the moment I don't have a good idea how to compare in this case, do you have any ideas?

Not sure. Given the added complexity, perhaps it's better to stick to only handling `==` and `!=` in any case for now, since this static check is only for optimisations.

mununki commented on jscomp/ml/typecore.ml:
```diff
@@ -259,7 +259,7 @@ let constant : Parsetree.constant -> (Asttypes.constant, error) result =
        with Failure _ -> Error (Literal_overflow "int64")
      end
-  | Pconst_integer (i,Some 'n') ->Ok (Const_bigint i)
+  | Pconst_integer (i,Some 'n') ->Ok (Const_bigint (Ext_string.remove_leading_zeros i))
```
I guess this is a sweet spot to handle the string value of bigint for following steps of compiling.

mununki commented:
> Not sure. Given the added complexity, perhaps it's better to stick to only handling `==` and `!=` in any case for now, since these static checks are only for optimisations.

I've removed the compare from the parmatch and added tests. https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/226108d1d0f90e08675cc684533bb3a1ac1c82ec

cristianoc commented:
This looks good to me.
@cometkim @zth would you also take a final look -- anything else to check?

In any case, one could describe this as experimental support, so it's clear that it's possible that one will fine more corner cases that need fixing.

zth commented:
We should mention it being experimental in the v11.1 release blog post, and docs of course. cc @fhammerschmidt

zth commented on jscomp/core/lam_convert.ml:
```diff
@@ -254,6 +254,19 @@ let lam_prim ~primitive:(p : Lambda.primitive) ~args loc : Lam.t =
+  | Paddbigint -> prim ~primitive:Paddbigint ~args loc
+  | Psubbigint -> prim ~primitive:Psubbigint ~args loc
+  | Pmulbigint -> prim ~primitive:Pmulbigint ~args loc
+  | Pdivbigint _is_safe (*FIXME*) -> prim ~primitive:Pdivbigint ~args loc
```
I assume these fixme annotations are from something copied?

zth reviewed (approved):
I've added a comment or two about questions that I don't think was resolved yet, but other than that, looking great!

mununki commented:
@zth @cristianoc I have two things on my hand.
1. remove optimization for comparing: As getting reviews, we can't find how to optimization yet. e.g. `0b111n > 0x7n`
2. safe div, mod for bigint: I'm going to implement it in the same way as safe div, mod for int.

I should be able to finalize this over the weekend.

mununki commented:
We don't have a way to compare between bigint hexadecimal `0x`, octal `0o`, binary `0b` yet. We need to be able to compare so we can support bigint literals for pattern matching. If we only supported decimal literals, we would be able to compare string length with or without `-` and compare with string comparisons. When considering the trade-offs between the two, I decided that it would be better for users if the initial version only supported decimals and supported bigint literals for pattern matching. What do you think? https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/e5c5601d26607dde94d2053816284a954c73d9bd

mununki commented:
I've added the implementation of checking division by zero and throw `Division_by_zero` like int type value. https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/5ebab2dd8368f899f1d054b351e72540bac74a8e

So, I have finished implementing what I had planned to implement.

cristianoc commented on jscomp/core/js_exp_make.ml:
```diff
@@ -1265,16 +1265,22 @@ let bigint_comp (cmp : Lam_compat.comparison) ?comment (e0: t) (e1: t) =
-  else
-    bigint_op ?comment Div e0 e1
+  match e1.expression_desc with
+  | Number (Bigint { i = i1 }) when i1 = "0" || i1 = "-0" -> 
```
Not sure about this change.
What if `e1` is a variable and not a constant: should't the runtime to a check for division by zero?

mununki replied:
Correct! How about this? https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/7d22f9cebc05449469976fefb24ea39652f940ea

cristianoc commented on jscomp/ml/translcore.ml:
```diff
@@ -350,6 +370,19 @@ let primitives_table =
+      ("%addbigint", Paddbigint);
+      ("%subbigint", Psubbigint);
+      ("%mulbigint", Pmulbigint);
+      ("%divbigint", Pdivbigint Safe);
```
Looks like there's no way to generate an unsafe bigint div, as `Pdivbigint` always receives `Safe`?
If that's true and intended, then I guess `Pdivbigint` can be simplified and the runtime does not need to take a parameter.

mununki replied:
Removed https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/649dc3e170b8316bcf5ef794dc6de69cec9699f2

cometkim commented on jscomp/runtime/caml_bigint.res:
```diff
@@ -0,0 +1,37 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. */
+
+let div = (x: bigint, y: bigint) =>
+  if y == 0n {
+    raise(Division_by_zero)
+  } else {
+    Caml_bigint_extern.div(x, y)
+  }
+
+let mod_ = (x: bigint, y: bigint) =>
+  if y == 0n {
+    raise(Division_by_zero)
+  } else {
+    Caml_bigint_extern.rem(x, y)
+  }
```
I'm not sure this is really necessary, JS bigint already throw its own error (`RangeError` with its message) on div/mod by zero.

So this just duplicates the original behavior, and shadows original error class.

mununki replied:
This is for catching the res exn as Division_by_zero.
```res
try {
  ...
} catch {
|  Division_by_zero => ...
}
```

cometkim replied:
It's effective for pure ReScript code for sure, but not for external interoperability cases.

The best way I think would be to enable more handling of JS built-in exceptions other than `Js.Exn.Error`, e.g. `Js.Exn.RangeError` or `Js.Exn.TypeError` but the current compiler does not support this.

If we plan to support it in the future, it may be better to postpone this. Detection is possible with `exception Js.Exn.Error`.

cometkim replied:
(It's definitely not a pretty. The only reason I claim this is that bigint is the JS primitive)

mununki replied:
Yes, it is 🥲 I was trying to match the existing implementation of int.
I don't know if it's possible, but if it becomes possible to bind JS built-in exception with a customizable tagged variant, it would be nice to refactor this.

cometkim commented on jscomp/build_tests/super_errors/expected/bigint_match_literal.res.expected:
```diff
@@ -0,0 +1,11 @@
+
+  [1;33mWarning number 11[0m
```
just todo note. maybe it will be better to disable ansi color for test snapshot, or to checkin both colored and not

cometkim reviewed (approved):
All others looks fine. Great job @mununki

mununki commented:
Sorry for rebasing during review, there was a conflict in the change log file.

cristianoc commented on jscomp/core/js_exp_make.ml:
```diff
@@ -1253,6 +1259,23 @@ let rec int32_band ?comment (e1 : J.expression) (e2 : J.expression) :
+let bigint_comp (cmp : Lam_compat.comparison) ?comment (e0: t) (e1: t) =
+  bin ?comment (Lam_compile_util.jsop_of_comp cmp) e0 e1
+
+let bigint_div ~checked ?comment (e0: t) (e1: t) =
```
looks like the option to disable checking divide by zero is really only hidden under some ancient internal-only configuration

at some point, it can be removed, perhaps in a later PR

cristianoc commented on jscomp/core/lam_analysis.ml:
```diff
@@ -27,6 +27,7 @@ let not_zero_constant (x : Lam_constant.t) =
   match x with
   | Const_int { i } -> i <> 0l
   | Const_int64 i -> i <> 0L
+  | Const_bigint i -> i <> "0"
```
is this normalised, or are `-0` and `00` also possible?

mununki replied:
`00` is normalized, but `-0`. I'll add another clause `i <> "-0"`

cristianoc replied:
How about instead using a representation with a boolean (sign), and normalised string. So there's no doubt of the meaning and less likely to introduce bugs in future.

mununki replied:
Ah, very good. I'll refactor accordingly.

cristianoc commented on jscomp/others/js_bigint.res:
```diff
@@ -1,3 +1,74 @@
+Js.Bigint.fromString("")
+
+/* returns 17n */
+Js.Bigint.fromString("0x11")
```
So this is not supported as literal, but can be constructed in this way via parsing right?

cristianoc commented on jscomp/others/js_bigint.res:
```diff
@@ -1,3 +1,74 @@
+@val
+/**
+Parses the given `string` into a `bigint` using JavaScript semantics. Return the
+number as a `bigint` if successfully parsed, `null`, `undefined`, `_NaN` otherwise.
```
not sure what `null`, `undefined`, `_NaN` refers to.

mununki replied:
It is invalid comment, I'll remove this part.

cristianoc commented on jscomp/others/js_bigint.res:
````diff
@@ -1,3 +1,74 @@
+
+```rescript
+/* returns 123n */
+Js.Bigint.fromString("123")
````
is this called `fromStringExn`?

mununki replied:
Oops, you're correct.

cristianoc commented on jscomp/others/js_bigint.res:
````diff
@@ -1,3 +1,74 @@
+Js.Bigint.fromString("0o11")
+```
+*/
+external fromStringExn: string => bigint = "BigInt"
````
What's the exception? It would be good to have an example to see how to catch it.

mununki replied:
JS BigInt constructor throws error such as `Uncaught SyntaxError` in case of passing a non-numeric value. I'll add an example .

cristianoc commented on jscomp/ml/bigint_utils.ml:
```diff
@@ -0,0 +1,71 @@
+  | _ ->
+    (* If both numbers are either negative or positive, compare their lengths. *)
+    let len0, len1 = (String.length s0, String.length s1) in
+    if len0 = len1 then String.compare s0 s1  (* If lengths are equal, compare the strings directly. *)
```
this seems wrong for negative numbers

mununki replied:
It seems fine to me. Can you elaborate the case wrong?
https://ocaml.org/play#code=bGV0IGNvbXBhcmUgczAgczEgPQogICgqIGNoZWNrIGlmIG5lZ2F0aXZlICopCiAgbGV0IGlzX25lZyBzID0gU3RyaW5nLmxlbmd0aCBzID4gMCAmJiBzLlswXSA9ICctJyBpbgogIG1hdGNoIChpc19uZWcgczAsIGlzX25lZyBzMSkgd2l0aAogIHwgKHRydWUsIGZhbHNlKSAtPiAtMSAgKCogSWYgb25seSBzMCBpcyBuZWdhdGl2ZSwgaXQncyBzbWFsbGVyLiAqKQogIHwgKGZhbHNlLCB0cnVlKSAtPiAxICAgKCogSWYgb25seSBzMSBpcyBuZWdhdGl2ZSwgczAgaXMgbGFyZ2VyLiAqKQogIHwgXyAtPgogICAgKCogSWYgYm90aCBudW1iZXJzIGFyZSBlaXRoZXIgbmVnYXRpdmUgb3IgcG9zaXRpdmUsIGNvbXBhcmUgdGhlaXIgbGVuZ3Rocy4gKikKICAgIGxldCBsZW4wLCBsZW4xID0gKFN0cmluZy5sZW5ndGggczAsIFN0cmluZy5sZW5ndGggczEpIGluCiAgICBpZiBsZW4wID0gbGVuMSB0aGVuIFN0cmluZy5jb21wYXJlIHMwIHMxICAoKiBJZiBsZW5ndGhzIGFyZSBlcXVhbCwgY29tcGFyZSB0aGUgc3RyaW5ncyBkaXJlY3RseS4gKikKICAgIGVsc2UgaWYgbGVuMCA%2BIGxlbjEgdGhlbiAKICAgICAgaWYgaXNfbmVnIHMwIHRoZW4gLTEgZWxzZSAxICAoKiBBIGxvbmdlciBzMCBtZWFucyBpdCdzIGxhcmdlciB1bmxlc3MgaXQncyBuZWdhdGl2ZS4gKikKICAgIGVsc2UgICgqIGxlbjAgPCBsZW4xICopCiAgICAgIGlmIGlzX25lZyBzMSB0aGVuIDEgZWxzZSAtMSAgKCogQSBsb25nZXIgczEgbWVhbnMgczAgaXMgc21hbGxlciB1bmxlc3MgczEgaXMgbmVnYXRpdmUuICopCgpsZXQgKCkgPQooKiBsZW4wIDwgbGVuMSAgKikKICBhc3NlcnQoKGNvbXBhcmUgIi0xIiAiLTEwIiA9IDEpKTsKICBhc3NlcnQoKGNvbXBhcmUgIjEiICItMTAiID0gMSkpOwogIGFzc2VydCgoY29tcGFyZSAiLTEiICIxMCIgPSAtMSkpOwooKiBsZW4wID4gbGVuMSAgKikKICBhc3NlcnQoKGNvbXBhcmUgIi0xMDAiICItMTAiID0gLTEpKTsKICBhc3NlcnQoKGNvbXBhcmUgIjEwMCIgIi0xMCIgPSAxKSk7CiAgYXNzZXJ0KChjb21wYXJlICItMTAwIiAiMTAiID0gLTEpKTsKKCogbGVuMCA9IGxlbjEgICopCiAgYXNzZXJ0KChjb21wYXJlICItMTAiICItMTAiID0gMCkpOwogIGFzc2VydCgoY29tcGFyZSAiMTAiICItMTAiID0gMSkpOwogIGFzc2VydCgoY29tcGFyZSAiLTEwIiAiMTAiID0gLTEpKTsKICBwcmludF9lbmRsaW5lICJ0ZXN0IHBhc3NpbmchIg%3D%3D

mununki replied:
Ah, I figure it out this case is wrong `assert((compare "-10" "-12" = 1))` I'll fix it.

mununki replied:
https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/84eecd7d365e5b0c119c9883ede8eb9611140b7e

cristianoc commented on jscomp/ml/bigint_utils.ml:
```diff
@@ -0,0 +1,71 @@
+  000n -> 0n
+  001n -> 1n
+  01_000_000n -> 1000000n
+  0x1 -> 0x1n
```
this is now a syntax error too right?

cristianoc commented on jscomp/ml/bigint_utils.ml:
```diff
@@ -0,0 +1,71 @@
+  001n -> 1n
+  01_000_000n -> 1000000n
+  0x1 -> 0x1n
+  0x001n -> 0x001n
```
and this

cristianoc commented on jscomp/frontend/external_ffi_types.ml:
```diff
@@ -300,6 +300,9 @@ let inline_int_primitive (i : int32) : string list =
   [""; to_string (Ffi_inline_const (Const_int64 i))]
 
+let inline_bigint_primitive (i : string) : string list =
+  [""; to_string (Ffi_inline_const (Const_bigint i))]
```
This does not check that the bigint is valid, or remove leading zero's.
Don't know if this is a problem, just observing the fact.

cristianoc replied:
One possibility is to do the check `is_numeric` at parsing time, instead of type checking time. And do the normalisation (remove zeros) whenever a `Const_bigint` is constructed.

mununki replied:
Very good. I'm going to try to refactor `Const_bigint bool * string` in AST type.

cristianoc commented on jscomp/core/js_op.ml:
```diff
@@ -126,13 +126,11 @@ type 'a access = Getter | Setter
   | Int of { i : int32; c : int option }
   | Uint of int32
-  | Bigint of bigint_lit
+  | Bigint of bool * string
```
Small style suggestion: `Bigint of { positive: bool; value: string }`

mununki replied:
Good, it seems not going to be unboxed, still okay?

cristianoc replied:
The difference is going to be tiny in practice.

cristianoc commented on jscomp/core/js_exp_make.ml:
```diff
@@ -807,7 +807,11 @@ let tag_type = function
-  | Bigint i -> bigint i
+  | Bigint i -> 
+    let open Bigint_utils in
+    let sign, i = i |> remove_leading_sign in
```
style/naming: perhaps:
`let {positive, value} = parse_bigint i`

cristianoc commented on jscomp/core/lam.ml:
```diff
@@ -506,7 +506,7 @@ let prim ~primitive:(prim : Lam_primitive.t) ~args loc : t =
           Lift.bool
             (Lam_compat.cmp_float cmp (float_of_string a) (float_of_string b))
-      | Pbigintcomp cmp, Const_bigint a, Const_bigint b -> default ()
+      | Pbigintcomp cmp, Const_bigint (_, a), Const_bigint (_, b) -> default ()
```
nit: here `a` and `b` are never used, I thought the compiler would give a warning

cristianoc commented on jscomp/ml/bigint_utils.ml:
```diff
@@ -1,3 +1,15 @@
+  if len = 0 then (false, str)
+  else
+    if is_neg str || is_pos str then (not (is_neg str), String.sub str 1 (len -1))
+    else (not (is_neg str), str)
```
nit: in this branch `is_neg` is impossible, so just `(true, str)`

cristianoc commented on jscomp/ml/parmatch.ml:
```diff
@@ -1106,8 +1106,8 @@ let build_other ext env : Typedtree.pattern = match env with
-      (function Tpat_constant(Const_bigint i) -> String.length i | _ -> assert false)
-      (function i -> Tpat_constant(Const_bigint (String.make i '*')))
+      (function Tpat_constant(Const_bigint (sign, i)) -> String.length (Bigint_utils.to_string sign i) | _ -> assert false)
+      (function i -> Tpat_constant(Const_bigint (true, (String.make i '*'))))
```
not sure what `(String.make i '*')` is doing here?
I would have guessed `string_of_int i` instead.
Not sure how to trigger this code path, perhaps with some example?

mununki replied:
Not sure. I guess this function is used in pattern matching for gadt partially. But logically I think we can change `(String.make i '*')` to `string_of_int i` here for less space complexity.

cristianoc commented on jscomp/syntax/src/res_core.ml:
```diff
@@ -812,6 +812,12 @@ let parseConstant p =
     match p.Parser.token with
     | Int {i; suffix} ->
+      (* Only decimal literal is allowed for bigint *)
+      if suffix = Some 'n' && not (Bigint_utils.is_numeric i) then
```
nit: perhaps `is_numeric` could be called `is_valid`?

cristianoc commented on jscomp/syntax/src/res_core.ml:
```diff
@@ -812,6 +812,12 @@ let parseConstant p =
+      if suffix = Some 'n' && not (Bigint_utils.is_numeric i) then
+        Parser.err p
+          (Diagnostics.message
+             "Invalid decimal literal. Only decimal literal is allowed for \
```
perhaps "Invalid bigint literal".

mununki commented:
Fixed as reviewed. https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/d82b4bc4d2045dc5e8e616c88ef6c83278abdbda

cristianoc reviewed (approved):
Looks good to go!
Anything missing?

mununki commented:
Nothing left.
Thank you for your good reviews!

zth commented:
Fantastic @mununki , well done! 👏
